File: November2021/C++/Interval_List_Intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        vector<vector<int>> res;
        int i=0,j=0;
        while(i<firstList.size() && j<secondList.size())
        {
            if(firstList[i][1]<=secondList[j][1] )
            {
                if( firstList[i][1]>=secondList[j][0])
                {
                    if(firstList[i][0]>=secondList[j][0])
                        res.push_back(firstList[i]);
                    else //if(secondList[j][0]>firstList[i][0])
                        res.push_back({secondList[j][0],firstList[i][1]});
                }
                i++;
            }
            else //if(firstList[i][1]>=secondList[j][1] )
            {
                if(secondList[j][1]>=firstList[i][0])
                {
                    if(firstList[i][0]>=secondList[j][0])
                         res.push_back({firstList[i][0],secondList[j][1]});
                    else //if(firstList[i][0]<secondList[j][0])
                        res.push_back(secondList[j]);
                }
                j++;
            }
        }
        return res;
    }
};

File: November2021/IntervalListIntersections.java
class Solution {
	// TC : O(n)
	// SC : O(n)
	public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {

		List<int[]> ans = new ArrayList<>();

		int fIndex = 0;
		int sIndex = 0;
		int mS, mE;
		while(fIndex < firstList.length && sIndex < secondList.length){
			mS = Math.max(firstList[fIndex][0], secondList[sIndex][0]);
			mE = Math.min(firstList[fIndex][1], secondList[sIndex][1]);

			if(mE >= mS)
				ans.add(new int[]{mS, mE});

			if(firstList[fIndex][1] == mE) fIndex++;
			if(secondList[sIndex][1] == mE) sIndex++;
		}

		return ans.toArray(new int[ans.size()][2]);
	}

}




